pkg/scheduler/usergroupmanagement/user.go

```diff
@@ -0,0 +1,57 @@
+package usergroupmanagement
+
+import (
+	"sync/atomic"
+
+	"github.com/apache/incubator-yunikorn-core/pkg/common/resources"
+)
+
+type User struct {
+	name	string // Name of the user
+	maxResources	*resources.Resource // Max Resource configured per user
+	maxApplications	int32 // Max Applications configured per user
+	runningApplications	*int32 // Running Applications
+	usedGroup	string
+}
+
+func NewUser(user string) *User {
+	return &User{
+		name:	user,
+	}
+}
+
+func (u *User) GetName() string {
+	return u.name
+}
+
+func (u *User) SetMaxApplications(maxApplications int32) {
+	u.maxApplications = maxApplications
+}
+
+func (u *User) IncRunningApplications() {
+	atomic.AddInt32(u.runningApplications, 1)
+}
+
+func (u *User) DecRunningApplications() {
+	atomic.AddInt32(u.runningApplications, -1)
+}
+
+func (u *User) CanRun() bool {
+	if atomic.LoadInt32(u.runningApplications) < u.maxApplications {
```

Comment:
This part should be double-checked, because it feels racy. 

By the time `LoadInt()` completes, a separate goroutine might come along and increase `runningApplications`. This feels more sound:

```
	current := atomic.LoadInt32(&runningApplications)
	if atomic.CompareAndSwapInt32(&runningApplications, current, current) < u.maxApplications {
		// runningApplications was definitely not incremented if we end up here
	}
```

It's easy to get this wrong. Also, I don't think we deliver to 32-bit targets, so I'd recommend `int64` instead.

If performance is not a concern (and this does not look like a hot code path), then an ordinary `Mutex` is probably better because it's easiser to reason about. Atomic stuff is too low level.

Reply:
Another approach that might work, if you incorporate the increment inside `CanRun()`:

```
func (u *User) CanRun() bool {
  for {
    current := atomic.LoadInt32(&u.runningApplications)
    if current < u.maxApplications {
      return false
    }

    if atomic.CompareAndSwapInt32(&u.runningApplications, current, current + 1) {
      return true
    }
    
   // increment failed, so we re-attempt in an infinite loop
  }
}
```

This seems to be OK, too, but again, too low level.

Reply:
I think LoadInt32 ensures that load happens in atomic way irrespective of the in-flight writes. 

Yes, it is low level but it is good enough to meet our needs. I think Mutex is more than what we need in this particular context. Also, atomic package has been used in some other places too in shim. Thoughts?

Reply:
I was thinking about this and I've become more confident that even my first suggestion is wrong. You can't separate the loading from the increment. The two operations are atomic, but their composition is not.

Consider the following state with multiple Goroutines running simultaneously:

maxApplications = 30
runningApplications = 29

**Scenario 1**

The first is incorrect because of:
Goruoutine 1:  loads current value ->  checks if 29 < 30 -> returns true -> increments runningApplications to 30
Goruoutine 2:  loads current value -> [execution suspended for a brief period] -> checks if 29 < 30 (current value is already 30) -> returns true -> increments runningApplications to 31

Goruoutine 2 should not return true, but it does.

**Scenario 2**

The first suggestion defends against this scenario 1, but not against this:
Goruoutine 1:  loads current value ->  checks if 29 < 30 -> returns true -> increments runningApplications to 30 or 31
Goruoutine 2:  loads current value ->  checks if 29 < 30 -> returns true -> increments runningApplications to 31 or 30

You can't have `CanRun()` and `IncRunningApplications()` in two separate calls. The check and increment must behave atomically. It would only work if the call site of these methods were written as atomic:
```
guard  *sync.Mutex
...
	if user != nil {
                guard.Lock()
		if user.CanRun() {
			user.IncRunningApplications()
		} else {
                ...
                }
                guard.Unlock()
```

But this makes the usage of atomics unnecessary.

Reply:
cc @wilfred-s @craigcondit